File: src/foam/nanos/auth/WebAuthService.java
package foam.nanos.auth;

import foam.core.X;
import foam.nanos.util.LRULinkedHashMap;
import javax.security.auth.login.LoginException;
import java.util.Map;

/**
 * Created by marcroopchand on 2017-06-27.
 */
public class WebAuthService
  extends CachedUserAndGroupAuthService
{
  /**
   * Map of { userId: X }
   *
   * This will be used for web clients logging in since they can't
   * marshall a context
   * */
  protected Map<String, X> loginMap = new LRULinkedHashMap<>(10000);

  public void webLogin(String userId, String password) {
    if ( userId == null || userId == "" ) return;
    if ( password == null || password == "" ) return;

    try {
      if ( ! loginMap.containsKey(userId) ) {
        X x = super.login(userId, password);
        loginMap.put(userId, x);
      }
    } catch (LoginException e) {
      e.printStackTrace();
    }
  }

  public Boolean check(String userId, java.security.Permission permission) {
    if ( userId == null || userId == "" || permission == null ) return false;

    if ( loginMap.containsKey(userId) ) {
      return super.check(loginMap.get(userId), permission);
    }

    return false;
  }

  public void updatePassword(String userId, String oldPassword, String newPassword) {
    if ( userId == null || userId == "" ) return;

    if ( loginMap.containsKey(userId) ) {
      super.updatePassword(loginMap.get(userId), oldPassword, newPassword);
    }
  }

  public void logout(String userId) {
    if ( userId == null || userId == "" ) return;

    if ( loginMap.containsKey(userId) ) {
      super.logout(loginMap.get(userId));
      loginMap.remove(userId);
    }
  }
}

File: src/foam/nanos/auth/WebAuthService.js
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.INTERFACE({
  package: 'foam.nanos.auth',
  name: 'WebAuthService',

  methods: [
    {
      name: 'generateChallenge',
      returns: 'Promise',
      javaReturns: 'String',
      javaThrows: [ 'javax.naming.AuthenticationException' ],
      args: [
        {
          name: 'userId',
          javaType: 'long'
        }
      ]
    },
    {
      name: 'challengedLogin',
      returns: 'Promise',
      javaReturns: 'foam.nanos.auth.User',
      javaThrows: [ 'javax.naming.AuthenticationException' ],
      args: [
        {
          name: 'userId',
          javaType: 'long'
        },
        {
          name: 'challenge',
          javaType: 'String'
        }
      ]
    },
    {
      name: 'login',
      returns: 'Promise',
      javaReturns: 'foam.nanos.auth.User',
      javaThrows: [ 'javax.naming.AuthenticationException' ],
      args: [
        {
          name: 'email',
          javaType: 'String'
        },
        {
          name: 'password',
          javaType: 'String'
        }
      ]
    },
    {
      name: 'check',
      javaReturns: 'Boolean',
      returns: 'Promise',
      args: [
        {
          name: 'userId',
          javaType: 'long'
        },
        {
          name: 'permission',
          javaType: 'foam.nanos.auth.Permission'
        }
      ]
    },
    {
      name: 'updatePassword',
      returns: 'Promise',
      javaReturns: 'void',
      javaThrows: [ 'javax.naming.AuthenticationException' ],
      args: [
        {
          name: 'userId',
          javaType: 'long'
        },
        {
          name: 'oldPassword',
          javaType: 'String'
        },
        {
          name: 'newPassword',
          javaType: 'String'
        }
      ]
    },
    {
      name: 'logout',
      javaReturns: 'void',
      returns: 'Promise',
      args: [
        {
          name: 'userId',
          javaType: 'long'
        }
      ]
    }
  ]
});
